### PR TITLE
Fix exp-time parse backtrack by saving if '[' starts Array or Dictionary

### PR DESCRIPTION
**Pull request by modelorganism**
Save the results of discovering whether a given '[' is the start of an Array or a Dictionary. Remembering this prevents `parseExprCollection()` from having to parse the first sub-expr in square brackets {2^n} times, where n is # of nested square brackets.

Swift in `parseExprCollection` parses the first expr in a `[x..]`, a square-bracketed expr-list, twice. Once to see if there is a colon after `x` or not, `[x : v ..]` vs, `[x , y ..]`, then a second parse 'for real'. If there is a `[..]` inside of the first expr of another `[..]`, `[[..]...]`,  the first expr of the inner list is parsed 4-times, or more generally 2^n, where n is the depth of nesting.

This PR adds a structure to class Parser, `ArrayOrDictStartLocCache`, which remembers whether each **[** starts a Array or Dict for as long any surrounding `[...]` is being parsed. For the duration of a `[...]` being parsed this cached value will be used in place a performing a subsequent speculative parse. There are still 2 parses of each head of a  `[h ...]`, but not 2^n. Once the outermost **]** is encountered, the information is discarded as the parser can longer backtrack over   the subexprs.

Tests ran successfully on Mac at this rev (that is with this change after Merge pull request #20500 from xedin/additional-tests-for-autoclosure…)
  
Resolves [SR-9220](https://bugs.swift.org/browse/SR-9220).

### CODE REVIEW

File: test/Parse/structure_no_overflow.swift [hunk 1]
```diff
@@ -35,3 +35,16 @@ let c = ({({({({ ({({({([ ({({({({ ({({({([  ({({({({ ({({({([ ({({({({ ({({({({
...
+// CHECK-NOT: error
+let d = [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+        [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+        [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+        [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+         1
+         ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]] ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
+        ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]] ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
+        ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]] ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
+        ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]] ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
```
Comment (dan-zheng) [inline test/Parse/structure_no_overflow.swift]: I wonder how quickly does this parse now?

**Comment by rintaro** [comment]
Thanks for tackling this!
This is clever idea. But, we want to eliminate backtracking here. Because parsing expression might have some side effects, and backtracking might not be able to cancel them. Not sure how it's bad, but still.

Previously, we weren't using backtracking:
https://github.com/apple/swift/blob/bb157a070ec6534e4b534456d208b03adc07704b/lib/Parse/ParseExpr.cpp#L3042-L3058
However, when we introduced libSyntax parsing, we started to use backtracking here because `parseList()` which is used in `parseExprDictionary()`/`parseExprArray()` could not insert pre-parsed expression to the list. But using `SyntaxParsingContext::popIf()` in `parseExprCollection()` and `SyntaxParsingContext::addSyntax()` in `parseExprDictionary()`/`parseExprArray()`, we can inject pre-parsed expression to the list.

@modelorganism do you want to try that?

File: test/Parse/structure_no_overflow.swift (hunk 1 above)
Comment (modelorganism) [reply test/Parse/structure_no_overflow.swift]: let l = [[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]
        0.09 real         0.06 user         0.02 sys
let l = [[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]
        0.11 real         0.07 user         0.02 sys
let l = [[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]
        0.13 real         0.10 user         0.02 sys
let l = [[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]
        0.18 real         0.14 user         0.03 sys
let l = [[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]
        0.27 real         0.23 user         0.03 sys
let l = [[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]
        0.47 real         0.42 user         0.04 sys
let l = [[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]
        0.86 real         0.78 user         0.06 sys
let l = [[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]
        1.69 real         1.50 user         0.11 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]
        3.30 real         2.94 user         0.21 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]
        6.47 real         5.85 user         0.39 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]
       13.47 real        12.44 user         0.80 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]]
       26.47 real        24.59 user         1.61 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]]]
       51.32 real        47.20 user         3.63 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]]]]
      103.45 real        95.47 user         7.31 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
      280.94 real       195.70 user        14.23 sys
let l = [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
        463.93 real       386.87 user        27.48 sys

Comment (dan-zheng) [reply test/Parse/structure_no_overflow.swift]: Huh, 463.93 is still pretty long. Maybe it's the best we can do.

**Pull request closed without merging**